File: app/src/main/java/com/example/draw4u/ui/dashboard/DashboardFragment.java
package com.example.draw4u.ui.dashboard;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.example.draw4u.DiaryInfo;
import com.example.draw4u.R;
import com.example.draw4u.ResultDiary;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.google.firebase.firestore.QuerySnapshot;

import java.util.ArrayList;

public class DashboardFragment extends Fragment {

    private RecyclerView mRecyclerView;
    private Button btn_search;
    public EditText searchWord;
    private  MyAdapter mAdapter;
    private RecyclerView.LayoutManager mLayoutManager;
    private ArrayList<ResultDiary> diaryInfos = new ArrayList<>();
    private SwipeRefreshLayout refreshLayout;
    FirebaseFirestore db = FirebaseFirestore.getInstance();
    FirebaseAuth mAuth = FirebaseAuth.getInstance();

    public View onCreateView(@NonNull LayoutInflater inflater,
            ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_dashboard, container, false);
        searchWord = (EditText)view.findViewById(R.id.searchWord);
        btn_search = (Button)view.findViewById(R.id.btn_search);
        mRecyclerView = (RecyclerView)view.findViewById(R.id.recyclerview);
        mRecyclerView.setHasFixedSize(true);
        mLayoutManager = new LinearLayoutManager(getActivity());
        mRecyclerView.setLayoutManager(mLayoutManager);
        mRecyclerView.scrollToPosition(0);
        mAdapter = new MyAdapter(diaryInfos);
        mRecyclerView.setAdapter(mAdapter);
        mRecyclerView.addItemDecoration(new RecyclerViewDecoration(-33));
        mRecyclerView.setItemAnimator(new DefaultItemAnimator());
        refreshLayout = (SwipeRefreshLayout) view.findViewById(R.id.refresh_layout);

        //일기 검색
        btn_search.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View v) {
                initDataset(searchWord.getText().toString().trim());
            }
        });//일기 검색

        //리스트 갱신
        refreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                diaryInfos.clear();
                searchWord.setText("");
                initDataset();
                refreshLayout.setRefreshing(false);
            }
        });//리스트 갱신

        return view;
    }


    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        diaryInfos.clear();
        initDataset();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

    }

    //시작 dataset
    private void initDataset() {//맨 처음 data set 설정
        db.collection(mAuth.getUid())
                .orderBy("date", Query.Direction.DESCENDING)
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<QuerySnapshot> task) {
                        if (task.isSuccessful()) {//data 가져오기
                            for (QueryDocumentSnapshot document : task.getResult()) {
                                DiaryInfo diaryInfo = document.toObject(DiaryInfo.class);
                                ResultDiary resultDiary = new ResultDiary(diaryInfo);
                                diaryInfos.add(resultDiary);
                            }
                            mAdapter.notifyDataSetChanged();//Dataset Update
                        } else {
                            Toast.makeText(getActivity(), "ERROR!",
                                    Toast.LENGTH_SHORT).show();
                        }
                    }
                });

    }

    //일기 검색시 dataset
    private void initDataset(String searchWord) {//검색해서 data set 초기화

        db.collection(mAuth.getUid())
                .orderBy("date", Query.Direction.DESCENDING)
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<QuerySnapshot> task) {
                        if (task.isSuccessful()) {//data 가져오기
                            diaryInfos.clear();
                            for (QueryDocumentSnapshot document : task.getResult()) {
                                if(document.get("diary").toString().contains(searchWord)){
                                    DiaryInfo diaryInfo = document.toObject(DiaryInfo.class);
                                    ResultDiary resultDiary = new ResultDiary(diaryInfo);
                                    diaryInfos.add(resultDiary);
                                }
                                else {
                                    continue;
                                }
                            }
                            mAdapter.notifyDataSetChanged();//Dataset Update
                        } else {
                            Toast.makeText(getActivity(), "ERROR!",
                                    Toast.LENGTH_SHORT).show();
                        }
                    }
                });

    }

}

File: app/src/main/java/com/example/draw4u/ui/dashboard/DashboardFragment.kt
package com.example.draw4u.ui.dashboard

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.draw4u.DiaryInfo
import com.example.draw4u.R
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.android.synthetic.main.fragment_dashboard.*
import kotlinx.android.synthetic.main.item.view.*

class DashboardFragment : Fragment() {

  private lateinit var dashboardViewModel: DashboardViewModel
  var firestore : FirebaseFirestore? = null
  var fbAuth : FirebaseAuth? = null

  override fun onCreateView(
      inflater: LayoutInflater,
      container: ViewGroup?,
      savedInstanceState: Bundle?
    ):  View? {
      dashboardViewModel =
      ViewModelProviders.of(this).get(DashboardViewModel::class.java)
      val root = inflater.inflate(R.layout.fragment_dashboard, container, false)
    // 파이어스토어 인스턴스 초기화
    firestore = FirebaseFirestore.getInstance()

    val search_word: EditText = root.findViewById(R.id.searchWord)
    val search_btn: Button = root.findViewById(R.id.btn_search)
    val recyclerview: RecyclerView = root.findViewById(R.id.recyclerview)
    recyclerview.adapter = RecyclerViewAdapter()
    recyclerview.layoutManager = LinearLayoutManager(activity)

    search_btn.setOnClickListener(){
      (recyclerview.adapter as RecyclerViewAdapter).search(search_word.getText().toString())
    }

    return root
    }

  inner class RecyclerViewAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    // Person 클래스 ArrayList 생성성
    var diaryinfo : ArrayList<DiaryInfo> = arrayListOf()

    init {  // telephoneBook의 문서를 불러온 뒤 Person으로 변환해 ArrayList에 담음
      firestore?.collection(fbAuth?.uid.toString())?.addSnapshotListener { querySnapshot, firebaseFirestoreException ->
        // ArrayList 비워줌
        diaryinfo.clear()

        for (snapshot in querySnapshot!!.documents) {
          var item = snapshot.toObject(DiaryInfo::class.java)
          diaryinfo.add(item!!)
        }
        notifyDataSetChanged()
      }
    }

    // xml파일을 inflate하여 ViewHolder를 생성
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
      var view = LayoutInflater.from(parent.context).inflate(R.layout.item, parent, false)
      return ViewHolder(view)
    }

    inner class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
    }

    // onCreateViewHolder에서 만든 view와 실제 데이터를 연결
    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
      var viewHolder = (holder as ViewHolder).itemView

      viewHolder.date.text = diaryinfo[position].date
      viewHolder.keywords.text =
        "#" + diaryinfo[position].keyword1 + " #" + diaryinfo[position].keyword2 + " #" + diaryinfo[position].keyword3
    }

    // 리사이클러뷰의 아이템 총 개수 반환
    override fun getItemCount(): Int {
      return diaryinfo.size
    }

    // 파이어스토어에서 데이터를 불러와서 검색어가 있는지 판단
    fun search(searchWord : String) {
      firestore?.collection(fbAuth?.uid.toString())
        ?.addSnapshotListener { querySnapshot, firebaseFirestoreException ->
        // ArrayList 비워줌
        diaryinfo.clear()

        for (snapshot in querySnapshot!!.documents) {
          if (snapshot.getString("diary")!!.contains(searchWord)) {
            var item = snapshot.toObject(DiaryInfo::class.java)
            diaryinfo.add(item!!)
          }
        }
        notifyDataSetChanged()
      }
    }
  }

}